cpp/basix/polynomials.cpp

```diff
@@ -0,0 +1,106 @@
+{
+  xt::xtensor<double, 2> r0 = chebyshev_interval(d, xt::view(x, xt::all(), 0));
+  xt::xtensor<double, 2> r1 = chebyshev_interval(d, xt::view(x, xt::all(), 1));
+  xt::xtensor<double, 2> out({x.shape(0), static_cast<std::size_t>(polyset::dim(
+                                              cell::type::quadrilateral, d))});
+  std::size_t n = 0;
+  for (std::size_t i = 0; i < r0.shape(1); ++i)
+    for (std::size_t j = 0; j < r1.shape(1); ++j)
+    {
+      for (std::size_t p = 0; p < x.shape(0); ++p)
+        out(p, n) = r0(p, i) * r1(p, j);
+      ++n;
+    }
+  assert(out.shape(1) == n);
+
+  return out;
+}
+//-----------------------------------------------------------------------------
+xt::xtensor<double, 2> chebyshev_hex(int d, const xt::xarray<double>& x)
+{
+  xt::xtensor<double, 2> r0 = chebyshev_interval(d, xt::view(x, xt::all(), 0));
+  xt::xtensor<double, 2> r1 = chebyshev_interval(d, xt::view(x, xt::all(), 1));
+  xt::xtensor<double, 2> r2 = chebyshev_interval(d, xt::view(x, xt::all(), 2));
+  xt::xtensor<double, 2> out({x.shape(0), static_cast<std::size_t>(polyset::dim(
+                                              cell::type::hexahedron, d))});
+  std::size_t n = 0;
+  for (std::size_t i = 0; i < r0.shape(1); ++i)
+    for (std::size_t j = 0; j < r1.shape(1); ++j)
+      for (std::size_t k = 0; k < r2.shape(1); ++k)
+      {
+        for (std::size_t p = 0; p < x.shape(0); ++p)
+          out(p, n) = r0(p, i) * r1(p, j) * r2(p, k);
+        ++n;
+      }
+  assert(out.shape(1) == n);
+
+  return out;
+}
+//-----------------------------------------------------------------------------
+xt::xtensor<double, 2> tabulate_chebyshev(cell::type celltype, int d,
+                                          const xt::xarray<double>& x)
+{
+  switch (celltype)
+  {
+  case cell::type::interval:
+    return chebyshev_interval(d, x);
+  case cell::type::quadrilateral:
+    return chebyshev_quad(d, x);
+  case cell::type::hexahedron:
+    return chebyshev_hex(d, x);
+  default:
+    throw std::runtime_error("not implemented yet");
+  }
+}
+//-----------------------------------------------------------------------------
+} // namespace
+//-----------------------------------------------------------------------------
+xt::xtensor<double, 2> polynomials::tabulate(polynomials::type polytype,
+                                             cell::type celltype, int d,
+                                             const xt::xarray<double>& x)
+{
+  switch (polytype)
+  {
+  case polynomials::type::legendre:
+    return xt::view(polyset::tabulate(celltype, d, 0, x), 0, xt::all(),
+                    xt::all());
+  case polynomials::type::chebyshev:
+    return tabulate_chebyshev(celltype, d, x);
+  default:
+    throw std::runtime_error("not implemented yet");
+  }
+}
+//-----------------------------------------------------------------------------
+int polynomials::dim(polynomials::type polytype, cell::type cell, int d)
+{
+  switch (polytype)
```

Comment:
Why is a switch needed here?

Reply:
I was envisaging needing a switch for some other polynomial types (eg bubble polynomials) but as they're not implemented yet, I've removed the switch for now